### PR TITLE
DEV: Remove gifsicle dependency

### PR DESCRIPTION
**Pull request by nbianca**

### CODE REVIEW

**Comment by xfalcox** [comment]
We should add a PSA to the announcements when we merge this.

**Comment by nbianca** [comment]
This is ready now. FastImage v2.2.0 contains my additions to the gem.

File: app/jobs/onceoff/rebake_gif_images.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module Jobs
+  class RebakeGifImages < ::Jobs::Onceoff
+    def execute_onceoff(args)
+      Upload.where("original_filename LIKE '%.gif'").find_each do |upload|
+        uri = Discourse.store.path_for(upload) || upload.url
+        upload.update!(animated: FastImage.animated?(uri))
```
Comment (ZogStriP) [inline app/jobs/onceoff/rebake_gif_images.rb:8]: Should we catch any exceptions here?

File: app/jobs/onceoff/rebake_gif_images.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      User
+        .joins(:uploaded_avatar)
+        .where("uploads.original_filename LIKE '%.gif'")
+        .find_each do |user|
+
+        if user.uploaded_avatar.animated
+          if path = Discourse.store.path_for(user.uploaded_avatar)
+            file = File.new(path)
+          else
+            find = Discourse.store.download(user.uploaded_avatar)
```
Comment (ZogStriP) [inline app/jobs/onceoff/rebake_gif_images.rb:20]: I think you meant `file` here?

```suggestion
            file = Discourse.store.download(user.uploaded_avatar)
```

File: app/jobs/onceoff/rebake_gif_images.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+          end
+
+          # We want to reupload the avatar, but we want to keep the avatar
+          # until the new one is ready, so we will have to keep it under a
+          # different hash (like we do for secure media).
+          user.uploaded_avatar.update!(sha1: SecureRandom.hex(20))
+
+          upload = UploadCreator.new(file, user.uploaded_avatar.original_filename, type: "avatar").create_for(user.id)
+          user.update!(uploaded_avatar: upload)
```
Comment (ZogStriP) [inline app/jobs/onceoff/rebake_gif_images.rb:29]: Should we catch any exceptions here?

File: app/jobs/onceoff/rebake_gif_images.rb (hunk 1 above)
Comment (nbianca) [reply app/jobs/onceoff/rebake_gif_images.rb:8]: I do not think so. What exceptions do you have in mind?

FastImage should not throw (there is an option for that, disabled by default) and if AR throws then there is something more serious going on and we should see that in logs. If a onceoff job throws, it seems that it is repeated until it succeeds, which  is desired behavior.

File: app/jobs/onceoff/rebake_gif_images.rb (hunk 2 above)
Comment (nbianca) [reply app/jobs/onceoff/rebake_gif_images.rb]: Yes, you are right. 🤦🏻‍♀️

File: app/jobs/onceoff/rebake_gif_images.rb (hunk 3 above)
Comment (nbianca) [reply app/jobs/onceoff/rebake_gif_images.rb]: Same as above.

**Review by ZogStriP — APPROVED** [review]
LGTM 👍 

Should definitely come with a PSA on meta as well 😉

**Comment by nbianca** [comment]
I simplified the logic for rebaking animated GIF images and moved it to a scheduled job (it was a once-off job before). The reason for this change is to split the load of downloading and parsing every GIF images over a longer time period. Also, it no longer recreates uploads as a simple recreation of the optimized images is enough (these are generally used).

File: app/jobs/scheduled/rebake_gif_images.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+module Jobs
+  class RebakeGifImages < ::Jobs::Scheduled
+    every 1.hour
+
+    MAX_PROCESSED_GIF_IMAGES ||= 200
+
+    def execute(args)
+      Upload
+        .where("original_filename LIKE '%.gif'")
```
Comment (SamSaffron) [inline app/jobs/scheduled/rebake_gif_images.rb:11]: Should we use `.where("extension = 'gif' || original_filename like '%.gif' ")` ... gifs can sneak in with the wrong filename we rely on actual content I think.

Comment (SamSaffron) [reply app/jobs/scheduled/rebake_gif_images.rb:11]: It can be null though in some cases

**Comment by SamSaffron** [comment]
hmmm if we are not using "animated" anywhere why bother backfilling now? I think once we actually use this column it makes sense to have a back filler ... but, is there any point doing this early?

File: app/jobs/scheduled/rebake_gif_images.rb (hunk 4 above)
Comment (nbianca) [reply app/jobs/scheduled/rebake_gif_images.rb]: That sounds like a good idea. :+1:

File: lib/tasks/uploads.rake [hunk 5]
```diff
@@ -1194,3 +1194,23 @@ task "uploads:fix_missing_s3" => :environment do
     end
   end
 end
+
+task "uploads:backfill_animated" => :environment do
+  RailsMultisite::ConnectionManagement.each_connection do |db|
+    puts "Backfilling #{db}..."
+    Upload
+      .where('original_filename LIKE \'%.gif\'')
```
[the remark this answers is not in the document]
Comment (nbianca) [reply lib/tasks/uploads.rake]: I removed the backfilling Rake task completely. We do not need one now to force the process and we already have the scheduled job anyway...

File: app/jobs/scheduled/update_animated_uploads.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    every 1.hour
+
+    MAX_PROCESSED_GIF_IMAGES ||= 200
+
+    def execute(args)
+      Upload
+        .where("extension = 'gif' OR (extension IS NULL AND original_filename LIKE '%.gif')")
+        .where(animated: nil)
+        .limit(MAX_PROCESSED_GIF_IMAGES)
+        .find do |upload|
```
Comment (ZogStriP) [inline app/jobs/scheduled/update_animated_uploads.rb:14]: Shouldn't that be `each` instead of `find`?

**Pull request merged**